allennlp/models/encoder_decoders/encoder_decoder.py

```diff
@@ -0,0 +1,320 @@
+           are also represented as a ``TextField``.
+        """
+        # (batch_size, input_sequence_length, encoder_output_dim)
+        embedded_input = self._source_embedder(source_tokens)
+        source_mask = get_text_field_mask(source_tokens)
+        encoder_outputs = self._encoder(embedded_input, source_mask)
+        final_encoder_output = encoder_outputs[:, -1]  # (batch_size, encoder_output_dim)
+        if target_tokens:
+            targets = target_tokens["tokens"]
+            target_sequence_length = targets.size()[1]
+            # The last input from the target is either padding or the end symbol. Either way, we don't have to
+            # process it.
+            num_decoding_steps = target_sequence_length - 1
+        else:
+            num_decoding_steps = self._max_decoding_steps
+        # TODO (pradeep): Define a DecoderState class?
+        decoder_hidden = final_encoder_output
+        with torch.cuda.device_of(decoder_hidden):
+            decoder_context = Variable(torch.zeros(1, self._decoder_output_dim))
+        last_predictions = None
+        step_logits = []
+        step_probabilities = []
+        step_predictions = []
+        for timestep in range(num_decoding_steps):
+            if self.training and all(torch.rand(1) >= self._scheduled_sampling_ratio):
+                input_choices = targets[:, timestep]
+            else:
+                if timestep == 0:
+                    # For the first timestep, when we do not have targets, we input start symbols.
+                    # (batch_size,)
+                    with torch.cuda.device_of(final_encoder_output):
+                        input_choices = Variable(torch.LongTensor([self._start_index]).expand_as(
+                                final_encoder_output[:, 0]))
+                else:
+                    input_choices = last_predictions
+            decoder_input = self._prepare_decode_step_input(input_choices, decoder_hidden,
+                                                            encoder_outputs, source_mask)
+            decoder_hidden, decoder_context = self._decoder_cell(decoder_input,
+                                                                 (decoder_hidden, decoder_context))
+            # (batch_size, num_classes)
+            output_projections = self._output_projection_layer(decoder_hidden)
+            # list of (batch_size, 1, num_classes)
+            step_logits.append(output_projections.unsqueeze(1))
+            decoder_predictions = self._predict_step(output_projections)
+            step_probabilities.append(decoder_predictions["class_probabilities"].unsqueeze(1))
+            last_predictions = decoder_predictions["predicted_classes"]
+            # (batch_size, 1)
+            step_predictions.append(last_predictions.unsqueeze(1))
+        # step_logits is a list containing tensors of shape (batch_size, 1, num_classes)
+        # This is (batch_size, num_decoding_steps-1, num_classes)
```

Comment:
Isn't it `(batch_size, num_decoding_steps, num_classes)`?  Where'd the `-1` come from?

Reply:
Yes, you're right. I'll fix it.